### DISCUSSION TITLE
Right to left support

### DISCUSSION THREAD

**Discussion by Felienne** [hedyorg/hedy#766] [Archive]
At this point, we do not yet support right to left languages like Arabic and Hebrew and clearly, that is something we should. It is not trivial, because it will involve:

* mirroring the whole website
* mirroring Ace
* Making sure the cursor moves the right way in Ace

This topic is meant to discuss the impact of the r2l feature and also to talk about things the core team (all left to right readers) might have missed.

**Comment by Felienne** [comment]
I have tried to understand whether Ace even supports r2l, and the answer is... I don't know. Here are some issues worth investigating:

* https://github.com/ajaxorg/ace/pull/3658
* https://github.com/ajaxorg/ace/pull/3137 (Discussion on monospacing)
* https://github.com/ajaxorg/ace/issues/3041 (Discussion on cursor/text direction)
* https://github.com/ajaxorg/ace/pull/3400 (More on text direction)

**Reply by cben** [reply]
FWIW https://chavascript.github.io/chavascript/ has an RTL Ace that seems to work well.
Most bidi customization is in https://github.com/ChavaScript/chavascript/blob/master/docs/src/index.js, in `buildEditor()` and `updateLineDirection()` functions.
(I have no idea if those are best practices — I don't know Ace — just remember coming across it and being curious how it worked :grin:.
  I speak Hebrew, ping me if you need help testing.)

**Reply by Felienne** [reply]
Thanks for the suggestion @cben! We have a draft PR open now with quite some progress: #1703 If you could do a bit of Hebrew translation so you can see if it all looks good, that'd be amazing!!

**Comment by Felienne** [comment]
Our web wizard @rix0rrr figured it out, we need an extra ace component, added in https://github.com/Felienne/hedy/pull/769!

**Comment by Felienne** [comment]
Here is some relevant info if we want to pursue this further: https://rtlstyling.com/posts/rtl-styling
